host/sampi/http.go

```diff
@@ -0,0 +1,211 @@
+	}
+	return hostSlice, nil
+}
+
+func (s *Cluster) AddJobs(req map[string][]*host.Job) (map[string]host.Host, error) {
+	s.state.Begin()
+	for host, jobs := range req {
+		if err := s.state.AddJobs(host, jobs); err != nil {
+			s.state.Rollback()
+			return nil, err
+		}
+	}
+	res := s.state.Commit()
+
+	for host, jobs := range req {
+		for _, job := range jobs {
+			s.state.SendJob(host, job)
+		}
+	}
+
+	return res, nil
+}
+
+// Host Service methods
+
+func (s *Cluster) RemoveJobs(hostID string, jobIDs []string) error {
+	s.state.Begin()
+	s.state.RemoveJobs(hostID, jobIDs...)
+	s.state.Commit()
+	return nil
+}
+
+func (s *Cluster) StreamHostEvents(ch chan host.HostEvent, done chan bool) error {
+	s.state.AddListener(ch)
+	go func() {
+		<-done
+		go func() {
+			// drain to prevent deadlock while removing the listener
+			for range ch {
+			}
+		}()
+		s.state.RemoveListener(ch)
+		close(ch)
+	}()
+	return nil
+}
+
+// HTTP Route Handles
+func listHosts(c *Cluster, w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	rh := httphelper.NewReponseHelper(w)
+	ret, err := c.ListHosts()
+	if err != nil {
+		rh.Error(err)
+		return
+	}
+	rh.JSON(200, ret)
+}
+
+func registerHost(c *Cluster, w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	rh := httphelper.NewReponseHelper(w)
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		rh.Error(err)
+		return
+	}
+
+	h := &host.Host{}
+	if err = json.Unmarshal(data, &h); err != nil {
+		rh.Error(err)
+		return
+	}
+	if h.ID == "" {
+		rh.Error(errors.New("sampi: host id must not be blank"))
+		return
+	}
+
+	ch := make(chan *host.Job)
+
+	c.state.Begin()
+	if c.state.HostExists(h.ID) {
+		c.state.Rollback()
+		rh.Error(errors.New("sampi: host exists"))
+		return
+	}
+	c.state.AddHost(h, ch)
+	c.state.Commit()
+	go c.state.sendEvent(h.ID, "add")
+
+	// "defer" cleanups in a goroutine that waits until the http stream is closed.
+	go func() {
+		<-w.(http.CloseNotifier).CloseNotify()
+		c.state.Begin()
+		c.state.RemoveHost(h.ID)
+		c.state.Commit()
+		go c.state.sendEvent(h.ID, "remove")
+	}()
+
+	wr := sse.NewSSEWriter(w)
+	enc := json.NewEncoder(wr)
+	w.Header().Set("Content-Type", "text/event-stream; charset=utf-8")
+	w.WriteHeader(200)
+	wr.Flush()
+	for data := range ch {
+		enc.Encode(data)
+		wr.Flush()
+	}
+}
+
+func addJobs(c *Cluster, w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	rh := httphelper.NewReponseHelper(w)
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		rh.Error(err)
+		return
+	}
+
+	var req map[string][]*host.Job
+	if err := json.Unmarshal(data, &req); err != nil {
+		rh.Error(err)
+		return
+	}
+	res, err := c.AddJobs(req)
+	if err != nil {
+		rh.Error(err)
+		return
+	}
+	rh.JSON(200, res)
+}
+
+func removeJob(c *Cluster, w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	hostID := ps.ByName("host_id")
+	jobIDs := []string{ps.ByName("job_id")}
+	if err := c.RemoveJobs(hostID, jobIDs); err != nil {
```

Comment:
`RemoveJobs` could have arguments `hostID, jobIDs ...string` to avoid the need for the single element slice above